File: src/main/java/chapter4/combine/ElectricBills.java
package chapter4.combine;

import common.Log;
import io.reactivex.Observable;
import org.apache.commons.lang3.tuple.Pair;

import java.text.DecimalFormat;

public class ElectricBills {

    private int index = 0;

    public void electricBillV1() {
        String[] data = {"100", "300"};

        Observable<Integer> basePrice = Observable.fromArray(data)
                .map(Integer::parseInt)
                .map(val -> {
                    if (val <= 200) {
                        return 910;
                    } else if (val <= 400) {
                        return 1600;
                    } else {
                        return 7300;
                    }
                });

        Observable<Integer> usagePrice = Observable.fromArray(data)
                .map(Integer::parseInt)
                .map(val -> {
                    double series1 = Math.min(200, val) * 93.3;
                    double series2 = Math.min(200, Math.max(val - 200, 0)) * 187.9;
                    double series3 = Math.min(0, Math.max(val - 400, 0)) * 280.65;
                    return (int) (series1 + series2 + series3);
                });

        Observable<Integer> source = Observable.zip(
                basePrice,
                usagePrice,
                (v1, v2) -> v1 + v2);

        source.map(val -> new DecimalFormat("#,###").format(val))
                .subscribe(val -> {
                    StringBuilder sb = new StringBuilder();
                    sb.append("Usage: " + data[index] + "kWh -> ");
                    sb.append("Price: " + val + "원");

                    Log.it(sb.toString());
                    index++;
                });
    }

    public void electricBillV2() {
        String[] data = {"100", "300"};

        Observable<Integer> basePrice = Observable.fromArray(data)
                .map(Integer::parseInt)
                .map(val -> {
                    if (val <= 200) {
                        return 910;
                    } else if (val <= 400) {
                        return 1600;
                    } else {
                        return 7300;
                    }
                });

        Observable<Integer> usagePrice = Observable.fromArray(data)
                .map(Integer::parseInt)
                .map(val -> {
                    double series1 = Math.min(200, val) * 93.3;
                    double series2 = Math.min(200, Math.max(val - 200, 0)) * 187.9;
                    double series3 = Math.min(0, Math.max(val - 400, 0)) * 280.65;
                    return (int) (series1 + series2 + series3);
                });

        Observable<Pair<String, Integer>> source = Observable.zip(
                basePrice,
                usagePrice,
                Observable.fromArray(data),
                (v1, v2, i) -> Pair.of(i, v1 + v2));

        source.map(val -> Pair.of(val.getLeft(),
                new DecimalFormat("#,###").format(val.getRight())))
                .subscribe(val -> {
                    StringBuilder sb = new StringBuilder();
                    sb.append("Usage: " + val.getLeft() + "kWh -> ");
                    sb.append("Price: " + val.getRight() + "원");

                    Log.it(sb.toString());
                });
    }

    public static void main(String[] args) {
        ElectricBills demo = new ElectricBills();
        demo.electricBillV1();
        demo.electricBillV2();
    }

}

File: src/main/kotlin/chapter4/combine/ElectricBills_.kt
package chapter4.combine

import common.Log
import io.reactivex.Observable
import io.reactivex.functions.BiFunction
import io.reactivex.functions.Function3
import java.text.DecimalFormat

class ElectricBills_ {

    private var index = 0

    fun electricBillV1() {
        val data = arrayOf("100", "300")

        val basePrice = Observable.fromArray(*data)
                .map { it.toInt() }
                .map {
                    when {
                        it <= 200 -> 910
                        it <= 400 -> 1600
                        else -> 7300
                    }
                }

        val usagePrice = Observable.fromArray(*data)
                .map { it.toInt() }
                .map {
                    val series1 = minOf(200, it) * 93.3
                    val series2 = minOf(200, maxOf(it - 200, 0)) * 187.9
                    val series3 = minOf(0, maxOf(it - 400, 0)) * 280.65
                    series1 + series2 + series3
                }
                .map { it.toInt() }

        val source = Observable.zip(basePrice, usagePrice,
                BiFunction { t1: Int, t2: Int -> t1 + t2 })

        source.map { DecimalFormat("#,###").format(it) }
                .subscribe {
                    val sb = StringBuilder()
                    sb.append("Usage: ${data[index]} kWh -> ")
                    sb.append("Price: ${it}원")
                    Log.it(sb.toString())

                    index++
                }
    }

    fun electricBillV2() {
        val data = arrayOf("100", "300")

        val basePrice = Observable.fromArray(*data)
                .map { it.toInt() }
                .map {
                    when {
                        it <= 200 -> 910
                        it <= 400 -> 1600
                        else -> 7300
                    }
                }

        val usagePrice = Observable.fromArray(*data)
                .map { it.toInt() }
                .map {
                    val series1 = minOf(200, it) * 93.3
                    val series2 = minOf(200, maxOf(it - 200, 0)) * 187.9
                    val series3 = minOf(0, maxOf(it - 400, 0)) * 280.65
                    series1 + series2 + series3
                }
                .map { it.toInt() }

        val source = Observable.zip(basePrice, usagePrice, Observable.fromArray(*data),
                Function3 { t1: Int, t2: Int, t3: String -> t3 to t1 + t2 })

        source.map { it.first to DecimalFormat("#,###").format(it.second) }
                .subscribe {
                    val sb = StringBuilder()
                    sb.append("Usage: ${it.first} kWh -> ")
                    sb.append("Price: ${it.second}원")
                    Log.it(sb.toString())
                }
    }

}


fun main(args: Array<String>) {
    val demo = ElectricBills_()
    demo.electricBillV1()
    demo.electricBillV2()
}
